src/Tasks/Microsoft.NET.Build.Tasks/GenerateShims.cs

```diff
@@ -0,0 +1,173 @@
+
+        /// <summary>
+        /// Path to project file (.csproj|.vbproj|.fsproj)
+        /// </summary>
+        [Required]
+        public string ProjectPath { get; set; }
+
+        /// <summary>
+        /// TFM to use for compile-time assets.
+        /// </summary>
+        [Required]
+        public string TargetFrameworkMoniker { get; set; }
+
+        /// <summary>
+        /// PackageId of the dotnet tool NuGet Package.
+        /// </summary>
+        [Required]
+        public string PackageId { get; set; }
+
+        /// <summary>
+        /// Package Version of the dotnet tool NuGet Package.
+        /// </summary>
+        [Required]
+        public string PackageVersion { get; set; }
+
+        /// <summary>
+        /// The command name of the dotnet tool.
+        /// </summary>
+        [Required]
+        public string ToolCommandName { get; set; }
+
+        /// <summary>
+        /// The entry point of the dotnet tool which will be run by Apphost
+        /// </summary>
+        [Required]
+        public string ToolEntryPoint { get; set; }
+
+        /// <summary>
+        /// The output Directroy path of generated shims.
+        /// </summary>
+        [Required]
+        public string PackagedShimOutputDirectory { get; set; }
+
+        /// <summary>
+        /// The RuntimeIdentifiers that shims will be generated for.
+        /// </summary>
+        [Required]
+        public ITaskItem[] PackageToolShimRuntimeIdentifiers { get; set; }
+
+        /// <summary>
+        /// Path of generated shims. metadata "ShimRuntimeIdentifier" is used to map back to PackageToolShimRuntimeIdentifiers.
+        /// </summary>
+        [Output]
+        public ITaskItem[] EmbeddedApphostPaths { get; private set; }
+
+        protected override void ExecuteCore()
+        {
+            NuGetFramework targetFramework = NuGetUtils.ParseFrameworkName(TargetFrameworkMoniker);
+            LockFile lockFile = new LockFileCache(this).GetLockFile(ProjectAssetsFile);
+            _packageResolver = NuGetPackageResolver.CreateResolver(lockFile, ProjectPath);
+
+            var embeddedApphostPaths = new List<ITaskItem>();
+            foreach (var runtimeIdentifier in PackageToolShimRuntimeIdentifiers.Select(r => r.ItemSpec))
+            {
+                var resolvedApphostAssetPath = GetApphostAsset(targetFramework, lockFile, runtimeIdentifier);
+
+                var packagedShimOutputDirectoryAndRid = Path.Combine(PackagedShimOutputDirectory, runtimeIdentifier);
+                var appHostDestinationFilePath =
+                    Path.Combine(packagedShimOutputDirectoryAndRid,
+                        $"{ToolCommandName}{Path.GetExtension(resolvedApphostAssetPath)}");
+                var appBinaryFilePath =
+                    $".store/{PackageId.ToLowerInvariant()}/{PackageVersion}/{PackageId.ToLowerInvariant()}/{PackageVersion}/tools/{targetFramework.GetShortFolderName()}/any/{ToolEntryPoint}";
```

Comment:
Strange to see this next to other Path.Combine calls. I think Path.Combine with each portion of path as an arg on its own line would be better.

Reply:
This is the embedded string. If we use Path.Combine, the embedded file will be different according to which machine builds it. \ build on win, / build on Linux. Although technically apphost is fine with that.

I will add comments and move it away from other Path.Combine

// This is the embedded string. We should normalize it on forward slash, so the file won't be different according to
// build machine.

Reply:
Will this looks better?

```c#

                // This is the embedded string. We should normalize it on forward slash, so the file won't be different according to
                // build machine.
                var appBinaryFilePath = string.Join('/',
                         ".store",
                         PackageId.ToLowerInvariant(),
                         PackageVersion,
                         PackageId.ToLowerInvariant(),
                         PackageVersion,
                         "tools",
                         targetFramework.GetShortFolderName(),
                         "any",
                         ToolEntryPoint);

```

Reply:
changed as above